src/nimony/typeprops.nim

```diff
@@ -0,0 +1,205 @@
+        skip field
+        while field.kind != ParRi:
+          let val = asEnumField(field).val
+          var thisVal: xint
+          case val.kind
+          of IntLit:
+            thisVal = createXint pool.integers[val.intId]
+          of UIntLit:
+            thisVal = createXint pool.uintegers[val.uintId]
+          else:
+            # enum field with non int/uint value?
+            return false
+          inc last
+          if thisVal != last:
+            return false
+          skip field
+      result = true
+    of DistinctT:
+      # check base type
+      var baseType = decl.body
+      inc baseType # skip distinct tag
+      result = isOrdinalType(baseType)
+    else:
+      result = isOrdinalType(decl.body)
+  of ParLe:
+    case typ.typeKind
+    of IntT, UIntT, CharT, BoolT:
+      result = true
+    of InvokeT:
+      # check base type
+      var base = typ
+      inc base # skip invoke tag
+      result = isOrdinalType(base)
+    else:
+      result = false
+  else:
+    result = false
+
+proc firstOrd*(c: var SemContext; typ: TypeCursor): xint =
+  case typ.kind
+  of Symbol:
+    let s = tryLoadSym(typ.symId)
+    if s.status != LacksNothing:
+      result = createNaN()
+      return
+    if s.decl.symKind != TypeY:
+      result = createNaN()
+      return
+    let decl = asTypeDecl(s.decl)
+    case decl.body.typeKind
+    of EnumT:
+      var field = asEnumDecl(decl.body).firstField
+      let firstVal = asEnumField(field).val
+      case firstVal.kind
+      of IntLit:
+        result = createXint pool.integers[firstVal.intId]
+      of UIntLit:
+        result = createXint pool.uintegers[firstVal.uintId]
+      else:
+        # enum field with non int/uint value?
+        result = createNaN()
+    of DistinctT:
+      # check base type
+      var baseType = decl.body
+      inc baseType # skip distinct tag
+      result = firstOrd(c, baseType)
+    else:
+      result = firstOrd(c, decl.body)
+  of ParLe:
+    case typ.typeKind
+    of IntT:
+      var bits = typ
+      inc bits # skip int tag
+      case typebits(c.g.config, bits.load)
+      of 8: result = createXint low(int8).int64
+      of 16: result = createXint low(int16).int64
+      of 32: result = createXint low(int32).int64
+      of 64: result = createXint low(int64)
+      else: result = createNaN()
+    of UIntT, CharT, BoolT:
+      result = zero()
+    of InvokeT:
+      # check base type
+      var base = typ
+      inc base # skip invoke tag
+      result = firstOrd(c, base)
+    else:
+      result = createNaN()
+  else:
+    result = createNaN()
+
+proc lastOrd*(c: var SemContext; typ: TypeCursor): xint =
+  case typ.kind
+  of Symbol:
+    let s = tryLoadSym(typ.symId)
+    if s.status != LacksNothing:
+      result = createNaN()
+      return
+    if s.decl.symKind != TypeY:
+      result = createNaN()
+      return
+    let decl = asTypeDecl(s.decl)
+    case decl.body.typeKind
+    of EnumT:
+      # check for holes
```

Comment:
Fine for now but we should just introduce `HolyEnumT` as that reflects reality better IMO.